### PR TITLE
Added init command line flag

### PR DESCRIPTION
**Pull request by JoshBrudnak**
I added a command line flag for creating a rustfmt configuration through an interactive mode. User is prompted for each basic rustfmt configuration. The choices are then saved in the `rustfmt.toml` file.

Fixes: #1490

### CODE REVIEW

File: src/bin/main.rs [hunk 1]
```diff
@@ -376,7 +380,7 @@ fn determine_operation(matches: &Matches) -> Result<Operation, ErrorKind> {
             return Ok(Operation::Help(HelpOp::None));
         } else if topic == Some("config".to_owned()) {
             return Ok(Operation::Help(HelpOp::Config));
-        } else if topic == Some("file-lines".to_owned()) {
+        } else if topic == Some("init".to_owned()) {
```
Comment (otavio) [inline src/bin/main.rs]: This seems unrelated, no?

File: src/bin/main.rs [hunk 2]
```diff
@@ -59,16 +59,14 @@ enum Operation {
...
-    // Print version information
+    /// Interactive mode for initializing the configuration for rustfmt
+    Init,
+    /// Print version information
     Version,
     /// Output default config to a file, or stdout if None
-    ConfigOutputDefault {
-        path: Option<String>,
-    },
+    ConfigOutputDefault { path: Option<String> },
```
Comment (otavio) [inline src/bin/main.rs]: Format changes, please make sure to be using the latest version of `rustfmt`.

File: src/bin/main.rs [hunk 3]
```diff
@@ -87,6 +85,11 @@ fn make_opts() -> Options {
         "Run in 'check' mode. Exits with 0 if input is formatted correctly. Exits \
          with 1 and prints a diff if formatting is required.",
     );
+    opts.optflag(
+        "",
+        "init",
+        "Setup rustfmt configuration in an interactive mode",
```
Comment (otavio) [inline src/bin/main.rs]: Setup or generate?

File: src/bin/main.rs [hunk 4]
```diff
@@ -380,6 +380,8 @@ fn determine_operation(matches: &Matches) -> Result<Operation, ErrorKind> {
             return Ok(Operation::Help(HelpOp::None));
         } else if topic == Some("config".to_owned()) {
             return Ok(Operation::Help(HelpOp::Config));
+        } else if topic == Some("file-lines".to_owned()) {
+            return Ok(Operation::Help(HelpOp::FileLines));
```
Comment (otavio) [inline src/bin/main.rs]: This should be squashed in previous commit.

File: src/config/interactive.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
```
Comment (otavio) [inline src/config/interactive.rs]: As this is a new file, I think 2018 is better for copyright.

File: src/config/interactive.rs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+/*
+impl<'a> iter::Iterator for InteractiveConfig<'a> {
+    type Item = &'a InteractiveConfig<'a>;
+
+    fn next(&mut self) -> Option<&InteractiveConfig<'a>> {
+        self.0.as_mut().and_then(Iterator::next)
+    }
+}
+*/
```
Comment (otavio) [inline src/config/interactive.rs]: Dead code.

File: src/config/interactive.rs [hunk 7]
```diff
@@ -47,6 +35,7 @@ impl<'a> InteractiveConfig<'a> {
     }
 }
 
+#[allow(dead_code)]
```
Comment (otavio) [inline src/config/interactive.rs]: Please don't.

**Comment by JoshBrudnak** [comment]
@otavio Should I prompt the user for configurations that are still only enabled in nightly since `--init` is only enabled in nightly rustfmt?

**Comment by otavio** [comment]
> @otavio Should I prompt the user for configurations that are only enabled in nightly since `--init` is only enabled in nightly rustfmt?

Well, this could be a question as well. So user decides it. What do you think?

**Comment by JoshBrudnak** [comment]
I think it might be better to determine whether the user is using nightly before prompting them for nightly configurations.

File: src/bin/main.rs (hunk 2 above)
Comment (otavio) [reply src/bin/main.rs]: Ah, use the version you are developing as it does change the format.

**Comment by otavio** [comment]
> I think it might be better to determine whether the user is using nightly before prompting them for nightly configurations.

Absolutely!

**Comment by otavio** [comment]
Is it possible for you to squash the commits? It allows for a good overview of the changes as a whole and to provide feedback. Once you do that, I will test it here as well :+1:

**Comment by JoshBrudnak** [comment]
> Is it possible for you to squash the commits? It allows for a good overview of the changes as a whole and to provide feedback. Once you do that, I will test it here as well

I squashed the previous 3 commits into the implementation commit.

**Comment by topecongiro** [comment]
@JoshBrudnak Hi, thank you for the PR! I really appreciate your effort and work on improving rustfmt.

However, I doubt whether this feature is useful. First of all, the original issue was about setting _configuration options_ through interactive mode, not _command line arguments_. Also non-user-facing configuration options should not be saved to rustfmt.toml.

**Comment by JoshBrudnak** [comment]
> However, I doubt whether this feature is useful. First of all, the original issue was about setting configuration options through interactive mode, not command line arguments. Also non-user-facing configuration options should not be saved to rustfmt.toml.

Would this feature be useful if I changed the questions to configure the configuration options? And if so which configuration options should I include? Thanks.

**Comment by otavio** [comment]
I support the idea of it providing a way to configure the formatting options in an interactive way as it allows for a very user-friendly way of showing the capabilities of the tool.

**Comment by nrc** [comment]
> I think it will be useful. I think you should only include options that is related to formatting. In particular, these options.
> Also, please make sure that on stable channel the list only includes stable options.

This all sounds good to me!

**Comment by topecongiro** [comment]
Closing due to inactivity. My apologies that the review was delayed for a long time.
